[Feature 6] Matching Algorithm

helenaglow opened the pull request:
**UPDATE: As per Hannah's advice, I have made changes!

The changes can be seen under the "Updated matching algorithm entities commit"! I am still planning on adding UserMatchesFactory interface and more in regard to the Factory design.

Rather than having the sorter, I am using comparable for the new class GroupScore. More information is added on the code.

Like before, there are still errors as I am under the assumption that getPublicProfile (for User) and getPreferences() for both User and Group will be added.

helenaglow committed:
Initial matching algorithm entities.

hannahpan1 commented on src/main/java/Entities/UserMatches/SelectionSortMatchGroups.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.HashMap;
+import java.util.List;
+
+public class SelectionSortMatchGroups implements Sorter{
```
For this I think we can use the comparable interface (since we are authoring the class we can use this instead of the Comparator interface) and the CompareTo instead of making our own sort interface. This will allow you to use the collections sort method as well as all other methods that can work with the Comparable interface. In comparison if we make our sort interface we can not access all the methods that know how to work with Comparable items. This would work if we have like the the match class implement comparable.

hannahpan1 commented on src/main/java/Entities/UserMatches/SelectionSortMatchGroups.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    //Sorts the groups based on their mapped values
+    public void sort(List<Object> groups) {
```
If we use the comparable interface then we won't have to specify the type of the list like List\<Object>, which seems to be an issue we are running into when you call sort on a list of groups instead of a list of objects (on line 34 of UserMatches.java).

hannahpan1 commented on src/main/java/Entities/UserMatches/UserMatches.java:
```diff
@@ -0,0 +1,77 @@
+
+    //Match the groups and Rank them!
+    @Override
+    public void match() {
```
Instead of having the match method put the similarity score and and group into a hash map, as mentioned before we could have a list of matches, where match is it's own class that has a similarity score instance attribute (this can be defined in the constructor the same way that calculateSimilarity() method is defining the similarity score). This match could implement comparable and then we can just make a list of match instances for each group in this method. Then we can just sort the list of matches using Collections' sort method. I think this will follow the clean architecture principle of single responsibility better and also get rid of some methods.

hannahpan1 reviewed (commented):
Great first implementation of your code! Wrote some comments about the implementation that I think might help simplify the code a bit.

helenaglow committed:
Updated matching algorithm entities.

ileuxali reviewed (commented):
I'll help make changes in user entity :^)

aaryavatsa reviewed (dismissed):
This looks good! Thank you for the nicely formatted documentation, it was very easy to follow.

helenaglow committed:
Matching Algorithm Use-Case (without screens and tests) and updated Matching Algorithm Entities

helenaglow commented:
I have uploaded the Matching Algorithm Use Case (without screens and tests) and updated Matching Algorithm Entities.

There some changes I will likely have to make depending on if we do serialization or csv.

Furthermore, at the moment there are two files I do not think I will need: MatchingAlgorithmInputBoundary and MatchingAlgorithmRequestModel. This is because the use case does not really have input. Currently I have booleans for whether the "Refresh" or "Find Group" buttons are clicked. However, I do not think this is necessary. Instead, in the screens, if those buttons are clicked the controller will be called. So, I do not think these files are necessary, but I will confirm with TA. 

Do you guys think I should make a UserMatches and GroupScore factory? I am not sure they are necessary, but I am making multiple instances of GroupScore. So I am not 100% sure.

**_I had another note but I cannot think of it at the moment. I will edit the message when I remember it :(_

helenaglow committed:
Merge branch 'main' of https://github.com/CSC207-2022F-UofT/course-project-groupedup into 2-feature-6-matching-algorithm

soheegoo commented on src/main/java/Entities/UserMatches/GroupScore.java:
```diff
@@ -0,0 +1,79 @@
+ * calculateSimilarity(): calculates and returns the similarity score based on the UserPublicProfile preferences
+ * of user and the preferences of group
+ */
+
```
Nice description! This makes it really easy to understand and follow your code. Thanks for making the such clear comments :)

aaryavatsa replied:
Agree with Sohee ^ setting the bar for doc strings 🥇  woooo

soheegoo commented on src/main/java/Entities/UserMatches/GroupScore.java:
```diff
@@ -0,0 +1,79 @@
+    public Double getScore(){
+        return this.score;
+    }
+
```
Really smart idea to implement the Comparable interface!

soheegoo commented on src/main/java/Entities/UserMatches/GroupScore.java:
```diff
@@ -0,0 +1,79 @@
+        return count / total;
+    }
+}
+
```
Maybe add some javadoc comments on how this function works to make it clear for the user

soheegoo commented on src/main/java/Entities/UserMatches/MatchingAlgorithm.java:
```diff
@@ -0,0 +1,7 @@
+    void match();
+
+}
+
```
Good use of clean architecture incase we want to introduce different matches in the future!

soheegoo commented on src/main/java/matching_algorithm_use_case/MatchingAlgorithmResponseModel.java:
```diff
@@ -0,0 +1,21 @@
+    List<String> groupToStringList;
+
+    public MatchingAlgorithmResponseModel(String uM, List<String> g){
+        this.updateMessage = uM;
```
Just a general note, I think you should write your input variables as like updateMatches to follow convention and make it easier for whoever is reading.

aaryavatsa reviewed (commented):
Everything looks good and I think you should be ready to merge -- but why is the autograder failing for this PR? It says there is a compliation error. Just glancing at it, I think it may be a problem with some of your method calls for the group leader. I think you should take a look at the autograder and the compilation error before we merge this.

helenaglow commented:
> Everything looks good and I think you should be ready to merge -- but why is the autograder failing for this PR? It says there is a compliation error. Just glancing at it, I think it may be a problem with some of your method calls for the group leader. I think you should take a look at the autograder and the compilation error before we merge this.

@aaryavatsa good catch it seems that there is an error with the group entity itself, but I didn't commit the group entity? (I don't use group leader anywhere in my code). 

 Also, there are is an error in group score as it is dependent on some methods and parts of Group and Group Public Profile that have not been added yet... not sure what I should do in this situation.

helenaglow committed:
[Feature 6] Matching Algorithm:

- I have added Java docs
- clarified the Request Model, will now be requesting User's username
- Changed UserMatches entites so that groups that the user is in or are not for the preferred course code are cut from the matches
- improved variable naming as Sohee suggested
- Presenter is now OutputBoundary

helenaglow committed:
[Feature 6] Matching Algorithm:

- I have added Java docs
- clarified the Request Model, will now be requesting User's username
- Changed UserMatches entites so that groups that the user is in or are not for the preferred course code are cut from the matches
- improved variable naming as Sohee suggested
- Presenter is now OutputBoundary

helenaglow committed:
[Feature 6] Matching Algorithm:

- I have added Java docs
- clarified the Request Model, will now be requesting User's username
- Changed UserMatches entites so that groups that the user is in or are not for the preferred course code are cut from the matches
- improved variable naming as Sohee suggested
- Presenter is now OutputBoundary

helenaglow commented:
I accidently kept committing without pushing. My bad.

Updates:
- I have added Java docs
- clarified the Request Model, will now be requesting User's username
- Changed UserMatches entites so that groups that the user is in or are not for the preferred course code are cut from the matches
- improved variable naming as Sohee suggested
- Presenter is now OutputBoundary

My code is still dependent on @soheegoo's .getGroupMembersUsernames() & .getGroupProfile(), and @hannahpan1's and @julia80887's .getPreferences() & .getCourseCode(). We already discussed and should be implemented with their coming pull requests

Furthermore, some more updates from my last pull request. I discussed with the TA at the lab about the implementation of factories. They said it was helpful when entities are rather complex, and it is up to me to decide what I deem to be necessary. I do not think a factory would be necessary for some reasons:

- UserMatches takes in an instance of User: this means that I would not have to make different cases of UserMatches based on the type of user. All match requirements are part of the User interface.
- I only make one UserMatches in the interactor
- It is not super complex


Please feel free to leave any comments or suggestions :)

hannahpan1 commented on src/main/java/Entities/UserMatches/MatchingAlgorithm.java:
```diff
@@ -1,5 +1,8 @@
+ * Interface for the Matching Algorithm just in case we need to implement it in a different way
+ */
 public interface MatchingAlgorithm{
     void match();
```
This is a great idea! We can also use this to add different ways to match items in the future.

hannahpan1 commented on src/main/java/Entities/UserMatches/UserMatches.java:
```diff
@@ -19,7 +19,7 @@
  *
  * METHOD DESCRIPTIONS:
  * match(): Will parse through the groups, create a GroupScore instance, and add it to matches. As GroupScore has a
```
Looks good! One thing to note is that in my most recent PR I have made course code an instance variable as well for the user public profile.  This means for cutByCourseCodeAndMemberList() method, instead of using the get method for preferences you can just look up course code.

hannahpan1 replied:
Also why do we get rid of group score in getMatches() when to turn them into a list of strings for the interactor we use getGroup anyways?

hannahpan1 commented on src/main/java/matching_algorithm_use_case/MatchingAlgorithmOutputBoundary.java:
```diff
@@ -5,7 +5,7 @@
+     * Send the response model (success message and ordered list) to the screens
      * @param responseModel: Update message and ordered List of Groups as Strings
      * @return an instance of the Matching Algorithm response model
      */
```
Does this run without error? The prepareFailView method is returning a MatchingAlgorithmResponseModel, which is not a string to my knowledge.

hannahpan1 reviewed (changes requested):
Hi! Everything looks good. I have a couple of questions I left in the review as well as one small change that I think needs to be implemented for our use cases to link together.

helenaglow committed:
-Updated .getCoursePreferences(), based on Hannah's pull request.
-Added getMatchesWithScore just in case we later decide to use the score

helenaglow committed:
- Spotted error in the attributes in interactor and fixed it.
- Aligned DsGateWay with Leo's pull request
- Updated output boundary java doc

helenaglow commented:
dsfdfsdfsf

helenaglow commented:
I have committed changes to the my entities based on Hannah's request and also updated my Interactor (which I found a small error in and a code smell 🤢). Also aligned my DsGateWay with Leo,s pull request

helenaglow committed:
Changed so that the Ordered list is converted into a string in UserMatches rather than the Interactor

ipekakyol reviewed (dismissed):
The javadocs make it really easy to follow the code. I do not see any clean architecture violations and everything looks good.

helenaglow committed:
Refactored packages to fit naming

hannahpan1 commented on src/main/java/entities/user_matches/GroupScore.java:
```diff
@@ -0,0 +1,88 @@
+import entities.Group;
+import entities.GroupProfile;
+import entities.User;
+import entities.UserPublicProfile;
```
Does this import statement work? To my knowledge our entities package is capitalized. This seems also to be the problem as to why your autograding is failing.

helenaglow committed:
Refactored Entity package to be capitalized for the time being. We will change it to be lowercase later on.

hannahpan1 reviewed (dismissed):
Looks good! Thanks for making all the new changes.

helenaglow committed:
Change getGroupProfile to getProfile based on Sohee's change

helenaglow committed:
Implemented Strategy Design Pattern and changed output boundary to return void
